text-minimessage/src/main/java/net/kyori/adventure/text/minimessage/tag/resolver/TagResolver.java

```diff
@@ -0,0 +1,403 @@
+import net.kyori.adventure.text.minimessage.tag.Tag;
+import net.kyori.adventure.text.minimessage.tag.builtin.BuiltInTags;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A collection of known tags.
+ *
+ * <p>A resolver can handle anywhere from a single tag, to a dynamically generated set of tags, returning a tag based on the provided name and arguments.</p>
+ *
+ * @see BuiltInTags
+ * @see Placeholder
+ * @since 4.10.0
+ */
+public interface TagResolver {
+  /**
+   * Create a new builder for a tag resolver.
+   *
+   * @return the tag resolver builder
+   * @since 4.10.0
+   */
+  static @NotNull Builder builder() {
+    return new TagResolverBuilderImpl();
+  }
+
+  /**
+   * Get the tag resolver that resolves all {@link BuiltInTags built-in tags}.
+   *
+   * <p>This is the default resolver used by parsers.</p>
+   *
+   * @return the default resolver
+   * @since 4.10.0
+   */
+  static @NotNull TagResolver standard() {
+    return BuiltInTags.all();
+  }
+
+  /**
+   * An empty tag resolver that will return {@code null} for all resolve attempts.
+   *
+   * @return the tag resolver
+   * @since 4.10.0
+   */
+  static @NotNull TagResolver empty() {
+    return EmptyTagResolver.INSTANCE;
+  }
+
+  /**
+   * A tag resolver that will resolve a single tag by a case-insensitive key.
+   *
+   * @param name the name of the tag to resolve
+   * @param tag the tag logic to return
+   * @return a new tag resolver
+   * @since 4.10.0
+   */
+  static TagResolver.@NotNull Single resolver(final @NotNull String name, final @NotNull Tag tag) {
+    SingleResolver.checkKey(name);
+    return new SingleResolver(
+      name,
+      requireNonNull(tag, "tag")
+    );
+  }
+
+  /**
+   * Create a tag resolver that only responds to a single tag name, and whose value does not depend on that name.
+   *
+   * @param name the name to respond to
+   * @param handler the tag handler, may throw {@link ParsingException} if provided arguments are in an invalid format
+   * @return a resolver that creates tags using the provided handler
+   * @since 4.10.0
+   */
+  static @NotNull TagResolver resolver(final @NotNull String name, final @NotNull BiFunction<ArgumentQueue, Context, Tag> handler) {
+    return resolver(Collections.singleton(name), handler);
+  }
+
+  /**
+   * Create a tag resolver that only responds to certain tag names, and whose value does not depend on that name.
+   *
+   * @param names the names to respond to
+   * @param handler the tag handler, may throw {@link ParsingException} if provided arguments are in an invalid format
+   * @return a resolver that creates tags using the provided handler
+   * @since 4.10.0
+   */
+  static @NotNull TagResolver resolver(final @NotNull Set<String> names, final @NotNull BiFunction<ArgumentQueue, Context, Tag> handler) {
```

Comment:
Would a `Predicate<String>` version of this method be useful?

Reply:
It might be *useful*, but one of my plans for later on is to be able to consolidate tag resolvers created using this method into a map lookup, rather than the iteration that `SequentialTagResolver` does currently, see #677. A `Predicate` method would prevent us from knowing all names ahead of time.

Since users can implement their own `TagResolver`s, I don't think it's necessary to add a Predicate method at the moment. I also doubt how useful it would be in this specific case, since the factory method here is specifically *not* passed the tag name. If it does turn into something people want in the future, that API could always be added at a later date.

Reply:
Sounds good to me!